zkevm-circuits/src/evm_circuit/execution/exp.rs

```diff
@@ -0,0 +1,304 @@
+    exponent_hi_is_zero: IsZeroGadget<F>,
+    /// Gadget to check if low 128-bit part of exponent is one or not.
+    exponent_lo_is_one: IsEqualGadget<F>,
+    /// Whether there is a single step in the exponentiation trace.
+    single_step: Cell<F>,
+    /// Gadget to check the byte-size of exponent.
+    exponent_byte_size: ByteSizeGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ExponentiationGadget<F> {
+    const NAME: &'static str = "EXP";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::EXP;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        // Query RLC-encoded values for base, exponent and exponentiation, where:
+        // base^exponent == exponentiation (mod 2^256).
+        let base_rlc = cb.query_rlc();
+        let exponent_rlc = cb.query_rlc();
+        let exponentiation_rlc = cb.query_rlc();
+
+        // Pop RLC-encoded base and exponent from the stack.
+        cb.stack_pop(base_rlc.expr());
+        cb.stack_pop(exponent_rlc.expr());
+
+        // Push RLC-encoded exponentiation to the stack.
+        cb.stack_push(exponentiation_rlc.expr());
+
+        // Extract low and high bytes of the base.
+        let (base_lo, base_hi) = (
+            from_bytes::expr(&base_rlc.cells[0x00..0x10]),
+            from_bytes::expr(&base_rlc.cells[0x10..0x20]),
+        );
+        // Extract low and high bytes of the exponent.
+        let (exponent_lo, exponent_hi) = (
+            from_bytes::expr(&exponent_rlc.cells[0x00..0x10]),
+            from_bytes::expr(&exponent_rlc.cells[0x10..0x20]),
+        );
+        // Extract low and high bytes of the exponentiation result.
+        let (exponentiation_lo, exponentiation_hi) = (
+            from_bytes::expr(&exponentiation_rlc.cells[0x00..0x10]),
+            from_bytes::expr(&exponentiation_rlc.cells[0x10..0x20]),
+        );
+
+        // We simplify constraints depending on whether or not the exponent is 0 or 1.
+        // In order to do this, we build some utility expressions.
+        let exponent_lo_is_zero = IsZeroGadget::construct(cb, exponent_lo.clone());
+        let exponent_hi_is_zero = IsZeroGadget::construct(cb, exponent_hi.clone());
+        let exponent_is_zero_expr =
+            and::expr([exponent_lo_is_zero.expr(), exponent_hi_is_zero.expr()]);
+        let exponent_lo_is_one = IsEqualGadget::construct(cb, exponent_lo.clone(), 1.expr());
+        let exponent_is_one_expr =
+            and::expr([exponent_lo_is_one.expr(), exponent_hi_is_zero.expr()]);
+
+        let zero_rlc = cb.query_word();
+        cb.require_zero(
+            "base * base + c == base^2 (c == 0)",
+            sum::expr(&zero_rlc.cells),
+        );
+        let base_sq = cb.query_word();
+
+        // If exponent == 0, base^exponent == 1, which implies:
+        // 1. Low bytes of exponentiation == 1
+        // 2. High bytes of exponentiation == 0
+        cb.condition(exponent_is_zero_expr.clone(), |cb| {
```

Comment:
Yes, good catch. I think I wish to stick to the `exponent_is_zero_expr`, i.e. using the `IsZeroGadget`s since those checks are stricter.

I'll update this PR: https://github.com/privacy-scaling-explorations/zkevm-specs/pull/300 with this change :)

Reply:
Made the change here: https://github.com/privacy-scaling-explorations/zkevm-specs/pull/300/commits/cb40782d9513a8fcdc3ff7722af800b098847316